shared/typeinference/codeql/typeinference/internal/TypeInference.qll

```diff
@@ -0,0 +1,776 @@
+        string toString();
+
+        /** Gets the location of this access. */
+        Location getLocation();
+
+        /**
+         * Gets the type at `path` for the type argument at position `tapos` of
+         * this access, if any.
+         *
+         * For example, in a method call like `M<int>()`, `int` is an explicit
+         * type argument at position `0`.
+         */
+        Type getTypeArgument(TypeArgumentPosition tapos, TypePath path);
+
+        /**
+         * Gets the inferred type at `path` for the position `apos` of this access.
+         *
+         * For example, if this access is the method call `M(42)`, then the inferred
+         * type at argument position `0` is `int`.
+         */
+        Type getInferredType(AccessPosition apos, TypePath path);
+
+        /** Gets the declaration that this access targets. */
+        Declaration getTarget();
+      }
+
+      /** Holds if `apos` and `dpos` match. */
+      bindingset[apos]
+      bindingset[dpos]
+      predicate accessDeclarationPositionMatch(AccessPosition apos, DeclarationPosition dpos);
+
+      /**
+       * Holds if matching an inferred type `t` at `path` inside an access at `apos`
+       * against the declaration `target` means that the type should be adjusted to
+       * `tAdj` at `pathAdj`.
+       *
+       * For example, in
+       *
+       * ```csharp
+       * void M(int? i) {}
+       * M(42);
+       * ```
+       *
+       * the inferred type of `42` is `int`, but it should be adjusted to `int?`
+       * when matching against `M`.
+       */
+      bindingset[apos, target, path, t]
+      default predicate adjustAccessType(
+        AccessPosition apos, Declaration target, TypePath path, Type t, TypePath pathAdj, Type tAdj
+      ) {
+        pathAdj = path and
+        tAdj = t
+      }
+    }
+
+    /**
+     * Provides logic for matching types at accesses against types at the
+     * declarations that the accesses target.
+     *
+     * Matching takes both base types and explicit type arguments into account.
+     */
+    module Matching<MatchingInputSig Input> {
+      private import Input
+
+      /**
+       * Holds if `a` targets `target` and the type for `apos` at `path` in `a`
+       * is `t` after adjustment by `target`.
+       */
+      pragma[nomagic]
+      private predicate adjustedAccessType(
+        Access a, AccessPosition apos, Declaration target, TypePath path, Type t
+      ) {
+        target = a.getTarget() and
+        exists(TypePath path0, Type t0 |
+          t0 = a.getInferredType(apos, path0) and
+          adjustAccessType(apos, target, path0, t0, path, t)
+        )
+      }
+
+      /**
+       * Gets the type of the type argument at `path` in `a` that corresponds to
+       * the type parameter `tp` in `target`.
+       */
+      bindingset[a, target]
+      pragma[inline_late]
+      private Type getTypeArgument(Access a, Declaration target, TypeParameter tp, TypePath path) {
```

Comment:
Would it make sense to add the context `a.getTarget() = target` to this predicate, and drop the inlining? It seems to depend on that context in all use-cases, and we'll materialize that combination anyway in `explicitTypeMatch`. That way the two negated occurrences would become simple anti-joins, and we could drop `explicitTypeMatch` (or rename this predicate to `explicitTypeMatch`).

Reply:
Makes sense to me.

Reply:
This will add more recursion, since `getTarget` is often mutually recursive with type inference, so that's why I did it like this. But it could deserve a comment saying so.